diff --git i/lib/k3/Core/MachineMessaging.k3 w/lib/k3/Core/MachineMessaging.k3
index ff11839..1dc5af1 100644
--- i/lib/k3/Core/MachineMessaging.k3
+++ w/lib/k3/Core/MachineMessaging.k3
@@ -205,7 +205,7 @@ control BroadcastByMachine[lbl : label, dest_trg : expr, masters : expr, masters
                          , send_ty : type, send_extra_fn : expr]
 {
   ( ignore (?e : collection ?t) ) @:Send
-    => (( ignore $.[e] ) @:SendBroadcast
+    => (( ignore $.[e] ) @:DirectSend
        ) @DispatchByMachine( lbl = lbl, dest_trg = dest_trg
                            , masters = masters, masters_map = masters_map
                            , send_ty = send_ty, send_extra_fn = send_extra_fn )
@@ -213,7 +213,7 @@ control BroadcastByMachine[lbl : label, dest_trg : expr, masters : expr, masters
     +> {}
 
   () @:Send
-    => ( ( ignore () ) @:SendBroadcast
+    => ( ( ignore () ) @:DirectSend
        ) @DispatchByMachine( lbl = lbl, dest_trg = dest_trg
                            , masters = masters, masters_map = masters_map
                            , send_ty = send_ty, send_extra_fn = send_extra_fn )
@@ -221,7 +221,7 @@ control BroadcastByMachine[lbl : label, dest_trg : expr, masters : expr, masters
     +> {}
 
   ( (\?x -> ?b) @:Receive )
-     => ( (\$#[x] -> $.[b]) @:Broadcast
+     => ( (\$#[x] -> $.[b]) @:DirectReceive
         ) @DispatchByMachine( lbl = lbl, dest_trg = dest_trg
                             , masters = masters, masters_map = masters_map
                             , send_ty = send_ty, send_extra_fn = send_extra_fn )
